torch/nn/parallel/distributed.py

```diff
@@ -442,10 +442,24 @@ def __init__(self, module, device_ids=None,
         # Whether to perform input tensor CPU to GPU copies on a side-stream
         self.use_side_stream_for_tensor_copies = os.environ.get("PYTORCH_DDP_USE_SIDE_STREAM", "1") == "1"
 
-        # Sync params and buffers
+        # Module replication within process (single-process multi device)
+        self._module_copies = self._replicate_modules_within_process()
+        # Build parameters for reducer.
+        parameters, expect_sparse_gradient = self._build_params_for_reducer()
+        # Verify model equivalence.
+        # Corresponding params' layouts (strides) must match across
+        # replicas within this process and across processes.
+        # see Note: "Gradient Layout Contract" in Reducer::initialize_buckets).
+        if self.device_ids is not None and len(self.device_ids) > 1:
+            dist._verify_replicas_within_process(parameters, expect_sparse_gradient)
+        self._verify_model_across_ranks(parameters)
+        # Sync params and buffers. Ensures all DDP models start off at the same value.
         self._sync_params_and_buffers(authoritative_rank=0)
```

Comment:
possibly we should still _sync_params_and_buffers first and then verify model consistency. it is possible that params are not exactly the same among ranks at the beginning, and this sync can make sure params are the same before verifying

Reply:
@zhaojuanmao Syncing params and buffers before verifying model replicas was the root cause of why this didn't work with Gloo before, though. In particular, `_sync_params_and_buffers` will ensure all params have the same starting values, but if parameter shapes are different, at least `gloo` will throw an error (not sure why NCCL gets past this, but might be due to the async nature of the collective).

On the other hand, `_verify_model_across_ranks` doesn't actually check the _values_ of the parameters, only that they have the same shape. So even if they have different values, that is fine, as we are only interested in ensuring all params have the same shape so that there's no hang during training, and the values will be filled in consistently after we do the shape check.

Reply:
I see, that makes sense!